src/Tidy/Codegen.purs

```diff
@@ -649,8 +649,17 @@ exprApp head =
 -- |     )
 -- | ```
 exprLambda :: forall e. Array (Binder e) -> Expr e -> Expr e
-exprLambda bnds body = bnds # NonEmptyArray.fromArray # maybe body \binders ->
-  ExprLambda { symbol: tokBackslash, binders, arrow: tokRightArrow, body }
+exprLambda bnds body = bnds # NonEmptyArray.fromArray # maybe body \bnds' ->
+  ExprLambda { symbol: tokBackslash, binders: addParens bnds', arrow: tokRightArrow, body }
+  where
+  addParens bnds' = bnds' <#> \b -> case b of
+    BinderTyped _ _ _ -> binderParens b
+    BinderOp _ _ -> binderParens b
+    BinderInt (Just _) _ _ -> binderParens b
+    BinderNumber (Just _) _ _ -> binderParens b
+    BinderConstructor _ vars | not $ Array.null vars -> binderParens b
+    _ -> b
```

Comment:
I think that this should be `precBinder2`, which should be fixed to use the same guard on `BinderConstructor`.

Reply:
I decided not to reuse `precBinder2` here or update it to use the same guard because it can be used in a `caseExpr`:
```purescript
case _ of
  NoParensHere a ->
-- vs
\(ParensHere a) ->
```

Reply:
OK, then we should move this to another `prec` function, and reuse it where appropriate. In the parser, this corresponds to all usages of `binderAtom` (https://github.com/purescript/purescript/blob/53b4117045f5517023f5d22f9fa506e651e4ff3e/lib/purescript-cst/src/Language/PureScript/CST/Parser.y#L575-L586). Seems like there's quite a few places where we should be using this precedence level.